Client/Helpers/TabEventHandler.swift

```diff
@@ -0,0 +1,147 @@
+ *
+ * ```
+ * extension HandoffHandler: TabEventHandler {
+ *     func tab(_ tab: Tab, didLoadPageMetadata metadata: PageMetadata) {
+ *         print("\(tab) has \(pageMetadata)")
+ *     }
+ *
+ *     func tab(_ tab: Tab, didLoadFavicon favicon: Favicon) {
+ *         print("\(tab) has \(favicon)")
+ *     }
+ * }
+ * ```
+ *
+ * Tab events should probably be only posted from one place, to avoid cycles.
+ *
+ * ```
+ * TabEvent.post(.didLoadPageMetadata(aPageMetadata), for: tab)
+ * ```
+ *
+ * In this manner, we are able to use the notification center and have type safety.
+ *
+ */
+// As we want more events we add more here.
+// Each event needs:
+// 1. a method in the TabEventHandler.
+// 2. a default implementation of the method – so not everyone needs to implement it
+// 3. a TabEventLabel, which is needed for registration
+// 4. a TabEvent, with whatever parameters are needed.
+//    i) a case to map the event to the event label (var label)
+//   ii) a case to map the event to the event handler (func handle:with:)
+protocol TabEventHandler {
+    func tab(_ tab: Tab, didLoadPageMetadata metadata: PageMetadata)
+    func tab(_ tab: Tab, didLoadFavicon favicon: Favicon)
+}
+
+// Provide default implmentations, because we don't want to litter the code with
+// empty methods, and `@objc optional` doesn't really work very well.
+extension TabEventHandler {
+    func tab(_ tab: Tab, didLoadPageMetadata metadata: PageMetadata) {}
+    func tab(_ tab: Tab, didLoadFavicon favicon: Favicon) {}
+}
+
+enum TabEventLabel: String {
+    case didLoadPageMetadata = "didLoadPageMetadata"
+    case didLoadFavicon = "didLoadFavicon"
+}
+
+enum TabEvent {
+    case didLoadPageMetadata(PageMetadata)
+    case didLoadFavicon(Favicon)
+    
+    var label: TabEventLabel {
+        switch self {
+        case .didLoadPageMetadata:
+            return .didLoadPageMetadata
+        case .didLoadFavicon:
+            return .didLoadFavicon
+        }
+    }
+    
+    func handle(_ tab: Tab, with handler: TabEventHandler) {
+        switch self {
+        case .didLoadPageMetadata(let metadata):
+            handler.tab(tab, didLoadPageMetadata: metadata)
+        case .didLoadFavicon(let favicon):
+            handler.tab(tab, didLoadFavicon: favicon)
+        }
+    }
+}
+
+// Hide some of the machinery away from the boiler plate above.
+////////////////////////////////////////////////////////////////////////////////////////
+extension TabEventLabel {
+    var name: Notification.Name {
+        return Notification.Name(self.rawValue)
+    }
+}
+
+extension TabEvent {
+    func notification(for tab: Any) -> Notification {
+        return Notification(name: label.name, object: tab, userInfo: ["payload": self])
+    }
+
+    /// Use this method to post notifications to any concerned listeners.
+    static func post(_ event: TabEvent, for tab: Any) {
+        NotificationCenter.default.post(event.notification(for: tab))
+    }
```

Comment:
Could this be generic?
```
protocol NotificationEvent {
 func notification(for object: Any) -> Notification {
   return Notification(name: label.name, object: object, userInfo: ["payload": self])
  }
  /// Use this method to post notifications to any concerned listeners.
  static func post(event: NotificationEvent, for object: Any) {
    NotificationCenter.default.post(event.notification(for: object))
  }
}
```

Reply:
It could be, but then it wouldn't be a `TabEvent`.

If we `post` it here with type `Any` we wouldn't be able to safely cast it to a `Tab` when we receive it.